src/engine/input/key-mapping/InputGroupList.cs

```diff
@@ -0,0 +1,180 @@
+    private InputEventWithModifiers latestDialogNewEvent;
+
+    public InputGroupList()
+    {
+        Instance = this;
+    }
+
+    public delegate void ControlSchemeChangedDelegate(Dictionary<string, List<InputEventWithModifiers>> data);
+
+    public event ControlSchemeChangedDelegate OnControlSchemeChanged;
+
+    public static InputGroupList Instance { get; private set; }
+    public IEnumerable<InputGroupItem> AllGroupItems => GetChildren().OfType<InputGroupItem>();
+
+    public static Dictionary<string, List<InputEventWithModifiers>> GetDefaultControlScheme()
+    {
+        return DefaultControlScheme;
+    }
+
+    public static Dictionary<string, List<InputEventWithModifiers>> GetCurrentlyAppliedControlScheme()
+    {
+        return InputMap.GetActions().OfType<string>()
+            .ToDictionary(p => p,
+                p => InputMap.GetActionList(p).OfType<InputEventWithModifiers>()
+                    .ToList());
+    }
+
+    public Dictionary<string, List<InputEventWithModifiers>> GetCurrentlyPendingControlScheme()
+    {
+        return AllGroupItems.SelectMany(p => p.Actions)
+            .ToDictionary(p => p.InputName, p => p.Inputs.Select(x => x.AssociatedEvent).ToList());
+    }
+
+    public override void _Ready()
+    {
+        conflictDialog = GetNode<ConfirmationDialog>(ConflictDialogPath);
+    }
+
+    public void ShowDialog(InputEventItem caller, InputEventItem conflict, InputEventWithModifiers newEvent)
+    {
+        latestDialogCaller = caller;
+        latestDialogConflict = conflict;
+        latestDialogNewEvent = newEvent;
+
+        conflictDialog.DialogText = $"There is a conflict with {conflict.AssociatedAction.DisplayName}.\n" +
+            $"Do you want to remove the input from {conflict.AssociatedAction.DisplayName}?";
+        conflictDialog.PopupCenteredMinsize();
+    }
+
+    /// <summary>
+    ///   Get the input conflict if there are any
+    /// </summary>
+    /// <param name="item">The event with the new value</param>
+    /// <returns>The collision if any</returns>
+    public InputEventItem Conflicts(InputEventItem item)
+    {
+        // Get all environments the item is associated with.
+        var environments = item.AssociatedAction.AssociatedGroup.Environments;
+
+        // Take all InputGroups.
+        // Take the ones with any interception of the environments.
+        // Take the input actions.
+        // Get the first action where the event collides or null if there aren't any.
+        return AllGroupItems.Where(p => p.Environments.Any(x => environments.Contains(x)))
+            .SelectMany(p => p.Actions)
+            .Where(p => !Equals(item.AssociatedAction, p))
+            .SelectMany(p => p.Inputs)
+            .FirstOrDefault(p => Equals(p, item));
+    }
+
+    public void InitFromData(Dictionary<string, List<InputEventWithModifiers>> data)
+    {
+        foreach (Node child in GetChildren())
+        {
+            RemoveChild(child);
```

Comment:
Not 100% but I think RemoveChild doesn't actually delete the child object, so this creates a bunch of dangling objects.

Reply:
I have looked at other calls to RemoveChild and found out about the Free method. Ill use it here